ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1505,62 +1508,40 @@ private module MkStage<StageSig PrevStage> {
+      Ap argAp, ApApprox argApa, Ap ap, ApApprox apa, Configuration config
     ) {
-      exists(ReturnKindExt kind |
-        fwdFlow(pragma[only_bind_into](ret), state, ccc,
-          TParamNodeSome(pragma[only_bind_into](summaryCtx.asNode())), apSome(argAp), ap, apa,
-          config) and
-        getApprox(argAp) = argApa and
-        kind = ret.getKind() and
-        parameterFlowThroughAllowed(summaryCtx, kind)
-      )
+      fwdFlow(pragma[only_bind_into](ret), state, ccc,
+        TParamNodeSome(pragma[only_bind_into](summaryCtx.asNode())),
+        pragma[only_bind_into](apSome(argAp)), ap, pragma[only_bind_into](apa),
+        pragma[only_bind_into](config)) and
+      kind = ret.getKind() and
+      parameterFlowThroughAllowed(summaryCtx, kind) and
+      argApa = getApprox(argAp) and
+      PrevStage::returnMayFlowThrough(ret, argApa, apa, kind, pragma[only_bind_into](config))
     }
 
     pragma[inline]
-    private predicate fwdFlowInMayFlowThrough(
-      DataFlowCall call, Cc cc, CcCall innerCc, ParamNodeOption summaryCtx, ApOption argAp,
-      ParamNodeEx param, Ap ap, ApApprox apa, Configuration config
-    ) {
-      fwdFlowIn(call, pragma[only_bind_into](param), _, cc, innerCc, summaryCtx, argAp, ap,
-        pragma[only_bind_into](apa), pragma[only_bind_into](config)) and
-      PrevStage::parameterMayFlowThrough(param, apa, config)
-    }
-
-    // dedup before joining with `flowThroughOutOfCall`
-    pragma[nomagic]
-    private predicate fwdFlowInMayFlowThroughProj(
-      DataFlowCall call, CcCall innerCc, ApApprox apa, Configuration config
-    ) {
-      fwdFlowInMayFlowThrough(call, _, innerCc, _, _, _, _, apa, config)
-    }
-
-    /**
-     * Same as `flowThroughOutOfCall`, but restricted to calls that are reached
-     * in the flow covered by `fwdFlow`, where data might flow through the target
-     * callable and back out at `call`.
-     */
-    pragma[nomagic]
-    private predicate fwdFlowThroughOutOfCall(
-      DataFlowCall call, CcCall ccc, RetNodeEx ret, NodeEx out, boolean allowsFieldFlow,
-      ApApprox argApa, ApApprox apa, Configuration config
+    private predicate fwdFlowThrough0(
+      DataFlowCall call, Cc cc, FlowState state, CcCall ccc, ParamNodeOption summaryCtx,
+      ApOption argAp, Ap ap, ApApprox apa, RetNodeEx ret, ReturnKindExt kind,
+      ParamNodeEx innerSummaryCtx, Ap innerArgAp, ApApprox innerArgApa, Configuration config
     ) {
-      fwdFlowInMayFlowThroughProj(call, ccc, argApa, config) and
-      flowThroughOutOfCall(call, ccc, ret, out, allowsFieldFlow, argApa, apa, config)
+      fwdFlowRetFromArg(pragma[only_bind_into](ret), kind, state, pragma[only_bind_into](ccc),
+        innerSummaryCtx, innerArgAp, innerArgApa, ap, pragma[only_bind_into](apa),
+        pragma[only_bind_into](config)) and
+      fwdFlowIsEntered(call, cc, ccc, summaryCtx, argAp, innerSummaryCtx, innerArgAp,
+        pragma[only_bind_into](config)) and
+      matchesCall(ccc, call)
```

Comment:
Isn't this implied by `fwdFlowIsEntered`? It ought to be.

Reply:
It is implied. `ccc = getContextCall(call, _, _)` implies `matchesCall(ccc, call)`.

Reply:
Fixed.